x/xcvm/keeper/transfer_intent.go

```diff
@@ -0,0 +1,448 @@
+	intentBeaconBlockHeader := *beaconBlockHeaders[0]
+	headerHash, err := intentBeaconBlockHeader.Hash()
+	if err != nil {
+		return fmt.Errorf("hash intent beacon block header: %v", err)
+	}
+
+	var currentParentRoot [32]byte
+	for _, header := range beaconBlockHeaders[1:] {
+		copy(currentParentRoot[:], header.GetParentRoot())
+		if currentParentRoot != headerHash {
+			return types.ErrInvalidBlockHeaders
+		}
+		headerHash, err = header.Hash()
+		if err != nil {
+			return fmt.Errorf("hash beacon block header: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// verifyBeaconBlockBody verifies that the provided beacon block body matches the beacon block header's body root and the transaction receipt's block hash
+func verifyBeaconBlockBody(beaconBlockHeader types.BeaconBlockHeader, beaconBlockBodySSZ []byte, txReceipt gethtypes.Receipt) error {
+	var beaconBlockBodyRoot [32]byte
+	beaconBlockBodyRootSlice := beaconBlockHeader.GetBodyRoot()
+	copy(beaconBlockBodyRoot[:], beaconBlockBodyRootSlice)
+
+	var beaconBlockBody prysmtypes.BeaconBlockBody
+	if err := beaconBlockBody.UnmarshalSSZ(beaconBlockBodySSZ); err != nil {
+		return fmt.Errorf("unmarshal beacon block body: %v", err)
+	}
+
+	beaconBlockBodyHash, err := beaconBlockBody.HashTreeRoot()
+	if err != nil {
+		return fmt.Errorf("hash beacon block body: %v", err)
+	}
+	if beaconBlockBodyHash != beaconBlockBodyRoot {
+		return types.ErrBlockBodyMismatch
+	}
+
+	blockHash := common.BytesToHash(beaconBlockBody.GetEth1Data().GetBlockHash())
+	if blockHash != txReceipt.BlockHash {
+		return types.ErrBlockHashMismatch
+	}
+
+	return nil
+}
+
+// getClientStateBeaconBlockHeader extracts the beacon block header from the client state
+func getClientStateBeaconBlockHeader(clientState ibccore.ClientState) (*types.BeaconBlockHeader, error) {
+	clientStateBz, err := proto.Marshal(clientState)
+	if err != nil {
+		return nil, fmt.Errorf("marshal client state: %v", err)
+	}
+	wasmClientState := new(wasmtypes.ClientState)
+	if err := proto.Unmarshal(clientStateBz, wasmClientState); err != nil {
+		return nil, fmt.Errorf("unmarshal client state bytes: %v", err)
+	}
+	typedClientState := new(codectypes.Any)
+	if err := typedClientState.Unmarshal(wasmClientState.Data); err != nil {
+		return nil, fmt.Errorf("unmarshal typed eth client state bytes: %v", err)
+	}
+	ethClientState := new(types.ClientState)
+	if err := ethClientState.Unmarshal(typedClientState.Value); err != nil {
+		return nil, fmt.Errorf("unmarshal eth client state bytes: %v", err)
+	}
+
+	return ethClientState.GetInner().GetFinalizedHeader(), nil
+}
+
+// verifyTransferEvent verifies that the ERC-20 transfer log emitted in the transaction receipt includes the correct destination address and amount specified in the transfer intent
+func verifyTransferEvent(txReceipt gethtypes.Receipt, intent types.TransferIntent, solverPublicKey *ecdsa.PublicKey) error {
+	type LogTransfer struct {
```

Comment:
is this part of the generated bindings? I feel we should have access to this type from elsehwere, but adding it here is no problem

Reply:
Yeah that's a good question. I wasn't able to find it in any bindings or as a geth struct. There is a [package](https://pkg.go.dev/github.com/brimstone/go-erc20#ERC20Transfer) specific for ERC20 transfer logic where we could reuse one of their structs, however I'm not sure if it's actively maintained and it also contains some extra information about the log that we don't care about (the only additional info we need is the token address). I'll look at ways to clean this up, may just end up moving it to a protobuf file for readability.